frame/airdrop/src/lib.rs

```diff
@@ -0,0 +1,1064 @@
+#![doc = include_str!("../README.md")]
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use pallet::*;
+
+pub mod models;
+pub mod weights;
+
+#[cfg(any(feature = "runtime-benchmarks", test))]
+mod benchmarking;
+mod mocks;
+#[cfg(test)]
+mod tests;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use crate::{
+		models::{Airdrop, AirdropState, Proof, RecipientFund, RemoteAccount},
+		weights::WeightInfo,
+	};
+	use codec::{Codec, FullCodec, MaxEncodedLen};
+	use composable_support::{
+		abstractions::{
+			nonce::{Increment, Nonce},
+			utils::{increment::SafeIncrement, start_at::ZeroInit},
+		},
+		math::safe::{SafeAdd, SafeSub},
+		types::{CosmosAddress, CosmosEcdsaSignature, EcdsaSignature, EthereumAddress},
+	};
+	use composable_traits::airdrop::AirdropManagement;
+	use frame_support::{
+		dispatch::PostDispatchInfo,
+		pallet_prelude::*,
+		traits::{
+			fungible::{Inspect, Transfer},
+			Time,
+		},
+		transactional, Blake2_128Concat, PalletId, Parameter,
+	};
+	use frame_system::pallet_prelude::*;
+	use p256::ecdsa::{signature::Verifier, Signature, VerifyingKey};
+	use scale_info::TypeInfo;
+	use sp_io::hashing::{keccak_256, sha2_256};
+	use sp_runtime::{
+		traits::{
+			AccountIdConversion, AtLeast32Bit, AtLeast32BitUnsigned, CheckedAdd, CheckedMul,
+			CheckedSub, Convert, One, Saturating, Verify, Zero,
+		},
+		AccountId32, DispatchErrorWithPostInfo, MultiSignature,
+	};
+	use sp_std::{fmt::Debug, vec::Vec};
+
+	/// [`AccountId`](frame_system::Config::AccountId) as configured by the pallet.
+	pub type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
+	/// [`AirdropId`](Config::AirdropId) as configured by the pallet.
+	pub type AirdropIdOf<T> = <T as Config>::AirdropId;
+	/// [`Airdrop`](crate::models::Airdrop) as configured by the pallet.
+	pub type AirdropOf<T> = Airdrop<
+		<T as frame_system::Config>::AccountId,
+		<T as Config>::Balance,
+		<T as Config>::Moment,
+	>;
+	/// [`Balance`](Config::Balance) as configured by the pallet.
+	pub type BalanceOf<T> = <T as Config>::Balance;
+	/// [`RecipientFund`](crate::models::RecipientFund) as configured by the pallet.
+	pub type RecipientFundOf<T> = RecipientFund<<T as Config>::Balance, <T as Config>::Moment>;
+	/// [`Moment`](Config::Moment) as configured by the pallet.
+	pub type MomentOf<T> = <T as Config>::Moment;
+	/// ['Proof'](crate::models::Proof) as configured by the pallet
+	pub type ProofOf<T> = Proof<<T as Config>::RelayChainAccountId>;
+	pub type RemoteAccountOf<T> = RemoteAccount<<T as Config>::RelayChainAccountId>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		AirdropCreated {
+			airdrop_id: T::AirdropId,
+			by: T::AccountId,
+		},
+		AirdropStarted {
+			airdrop_id: T::AirdropId,
+			at: T::Moment,
+		},
+		AirdropEnded {
+			airdrop_id: T::AirdropId,
+			at: T::Moment,
```

Comment:
Nit: Could include the unclaimed amount

Reply:
I'd rather calculate this as needed unless there is a performance gain from this. I suppose a function could exist to simplify the process - will do that.